### ISSUE TITLE
Logstash date filter fails to parse certain valid dates

### ISSUE THREAD

**Issue by gm310509** [elastic/logstash#10994]
While trying to index some tweets, I noted that certain date values do not parse using the "date" filter if a "target" is specified.

This applies to logstash 7.2.0 and elasticsearch 7.2.0 (buildhash 508c38a)

I have created an index as follows:
```JSON
PUT /auspol
{
  "mappings": {
    "properties": {
      "location": {
        "type": "geo_point"
      },
      "user_created_at": {
        "type": "date"
      },
      "retweet_count": {
        "type": "integer"
      }
    }
  }
}
```

My logstash filter is defined to include the following (full file is attached)
```
  date {
    match => [
      "user_created_at", "yyyy-MM-dd HH:mm:ss"
    ]
    target => "user_created_at"
    tag_on_failure => [ "user_created_at_parse_date_failure"]
  }
```
If I process `2010-10-03 02:15:54`  against this definition, I receive the following error and the document is not indexed:

`[WARN ] 2019-07-25 15:03:04.090 [[main]>worker0] elasticsearch - Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"auspol", :_type=>"_doc", :routing=>nil}, #<LogStash::Event:0x5d8034f5>], :response=>{"index"=>{"_index"=>"auspol", "_type"=>"_doc", "_id"=>"enGCJ2wBjjUvslPd3oLP", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"failed to parse field [user_created_at] of type [date] in document with id 'enGCJ2wBjjUvslPd3oLP'", "caused_by"=>{"type"=>"illegal_argument_exception", "reason"=>"failed to parse date field [2010-10-03 02:15:54] with format [strict_date_optional_time||epoch_millis]", "caused_by"=>{"type"=>"date_time_parse_exception", "reason"=>"Failed to parse with all enclosed parsers"}}}}}}`

**Strangely** if I supply the exact same date to the logstash filter below (yes it is a different source field name and no target is specified) it is parsed successfully and the document is indexed.
This can be seen in the final record in the sample data.

```
  date {
    match => [
      "created_at", "yyyy-MM-dd HH:mm:ss"
    ]
    remove_field => [ "created_at" ]
  }
```

If I change the failing date literal (in field user_created_at) by 1 year, 1 month, 1 day or 1 hour it parses successfully.  These attempts can also be seen in the sample data.

FWIW, the Sample data was created by taking a failing record from the full load job, replicating it and changing one value each time (except the final record which has two changes being the first field: created-at and the problem field: user_created_at).

I noted that someone else experienced issues with dates occurring at daylight savings transition times. So I tried setting my timezone (in the datefilter) to UTC, but still it failed to parse. Also, if this problem is related to daylight savings transitions, why does the second date filter definition (without the target specification) **not fail**?

The files attached are the full index definition, the full logstash configuration file and some sample data. The first record (excluding the header) is the one that fails to index.

[Index Definition](https://github.com/elastic/logstash/files/3429846/auspol-elasticsearch-admin-requests-denormalisedModel.txt)

[Logstash config file](https://github.com/elastic/logstash/files/3429849/auspol-test.conf.txt)

[Sample Data](https://github.com/elastic/logstash/files/3429851/auspol-test.csv.txt)

FWIW, I have tried parsing the same dates using the following SimpleDateFormat in Java and they parse successfully.
```Java
SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss");
```

**Comment by guyboertje** [comment]
As noted in our issue template, GitHub issues on this project are not the place for usage or debugging questions. Our forums reach a much wider audience.

I am closing this issue.

FWIW, I can't recreate the problem. Try adding `TZ=UTC` in front of the command to launch LS.

**Comment by yaauie** [comment]
FWIW, the rejection is coming from Elasticsearch, indicating that the string being specified cannot be parsed by either of the field's configured formats. The `strict_date_optional_time` format on the field is _close_ to matching, but requires a `T` between the date and time portions (see: [Joda Time format specification](http://joda-time.sourceforge.net/apidocs/org/joda/time/format/ISODateTimeFormat.html#dateOptionalTimeParser())). You may want to update the list of formats for the field in Elasticsearch indices and/or mapping templates.

**Comment by gm310509** [comment]
That is a disappointing resolution because I still think there is a bug (I am not qualified to determine exactly where - but I am experiencing the problem in logstash so I reported it here - I am perfectly comfortable with this being transferred to elasticsearch if the experts deem that is the correct place for it). I am not actually asking for assistance as I have another way of dealing with this.

Why do I believe this is a bug? Because if I put the exact same "problem date" into two different fields in a single data file and process them using two slightly different date filter specifications (shown below), one will parse and one will fail - the date formats are identical and the value is identical, only the filter specification is different (but not in a way that should be expected to affect the parse operation).
So it seems clear that there is a difference between how the exact same data is being processed by the parser where the exact same format is specified.

FWIW, I wasn't even going to log a bug until I discovered the difference in how the problem date was handled by these two different filter specifications.
 
Specifically, If I supply the problem date to this:
```
  date {
    match => [
      "user_created_at", "yyyy-MM-dd HH:mm:ss"
    ]
    target => "user_created_at"
    tag_on_failure => [ "user_created_at_parse_date_failure"]
  }
```
The error is generated.

Whereas if the problem date is supplied to this via another document in the same incoming data file (i.e. the environmental conditions are the same):
```
   date {
    match => [
      "created_at", "yyyy-MM-dd HH:mm:ss"
    ]
    remove_field => [ "created_at" ]
  }
```

No error is generated, the date is parsed and the document indexed.

I hope you will reconsider your choice to close this.

**Comment by gm310509** [comment]
> As noted in our issue template, GitHub issues on this project are not the place for usage or debugging questions. Our forums reach a much wider audience.
> 
> I am closing this issue.
> 
> FWIW, I can't recreate the problem. Try adding `TZ=UTC` in front of the command to launch LS.


@guyboertje I tried re-running as you suggested (i.e. set the timezone prior to running logstash). When I did that, the entire input file was successfully parsed and indexed.

However, there is still a bug of some sort IMHO. After this successful run, I reset the TZ to default (i.e. `unset TZ`) and recreated the problem. I then modified my date filter to include the timezone as follows:
```
  date {
    match => [
      "user_created_at", "yyyy-MM-dd HH:mm:ss"
    ]
    target => "user_created_at"
    tag_on_failure => [ "user_created_at_parse_date_failure"]
    timezone => "UTC"
  }
```
And I then re-encountered the error re the unparsable date. The error is occurring on *both* records with the invalid date.

It appears that the [logstash date filter timezone option](https://www.elastic.co/guide/en/logstash/current/plugins-filters-date.html#plugins-filters-date-timezone) is being ignored.

FWIW, I also had to add timezone => "UTC" to the other date filter.

```
[WARN ] 2019-07-26 09:56:42.898 [[main]>worker0] elasticsearch - Could not index event to Elasticsearch. {:status=>400, :action=>["index", {:_id=>nil, :_index=>"auspol", :_type=>"_doc", :routing=>nil}, #<LogStash::Event:0x53173585>], :response=>{"index"=>{"_index"=>"auspol", "_type"=>"_doc", "_id"=>"L42QK2wBuFSde8P8wTBa", "status"=>400, "error"=>{"type"=>"mapper_parsing_exception", "reason"=>"failed to parse field [user_created_at] of type [date] in document with id 'L42QK2wBuFSde8P8wTBa'", "caused_by"=>{"type"=>"illegal_argument_exception", "reason"=>"failed to parse date field [2010-10-03 02:15:54] with format [strict_date_optional_time||epoch_millis]", "caused_by"=>{"type"=>"date_time_parse_exception", "reason"=>"Failed to parse with all enclosed parsers"}}}}}}
```
IMHO, there is still a bug somewhere here as the timezone option in the date filter is seemingly being ignored. Also, why would setting the TZ variable in my terminal session running logstash affect the parsing of the date on the elasticsearch server (which is presumably running in the system default timezone)?

Since this appears to be Timezone related, the default timezone for the system is `DEFAULT_TIMEZONE="US/Eastern"`

**Comment by bidhanahdib** [comment]
I'm also having the same. Is there any soln for that?

**Comment by gm310509** [comment]
This was quite some time ago and I no longer have access to the system.
What I think I eventually worked out that it was a timezone mismatch issue. If memory serves (and it was a long time ago), I think I was receiving some UTC timestamps but somewhere in the pipeline they were being treated as local timestamps.
The problem was that when my local timezone changed to Daylight savings (changed back from Daylight savings?), there is an hour that is missing in real life. That is the local time jumps from 1:59:59 to 3:00:00 (for example). When I tried to ingest timestamps on that day that had times in the 2AM frame, they were being rejected.
Unfortunately, I do not remember where this was occurring (because it was so long ago), or what setting I had to change to make it work (because I no longer have access to the source), but that was the basis of the problem as I remember it.
